File: Server_Root/Server/RoomServer/PlayerInput/PlayerInputs/Rotation.cpp
#include "stdafx.h"
#include "Rotation.h"
#include "../../../FBSFiles/PlayerInput/Rotation_generated.h"

namespace PlayerInput
{
	Rotation::Rotation()
	{
	}

	Rotation::~Rotation()
	{
	}

	FBS::PlayerInputType Rotation::GetType()
	{
		return FBS::PlayerInputType::PlayerInputType_Rotation;
	}

	IPlayerInput* Rotation::Clone()
	{
		Rotation* pClone = new Rotation();

		pClone->m_nEntityID = m_nEntityID;
		pClone->m_vec3Rotation = m_vec3Rotation;

		return pClone;
	}

	const char* Rotation::Serialize(int* pLength)
	{
		FBS::Data::Vector3 rotation(m_vec3Rotation.x(), m_vec3Rotation.y(), m_vec3Rotation.z());

		FBS::PlayerInput::RotationBuilder data_builder(m_Builder);
		data_builder.add_EntityID(m_nEntityID);
		data_builder.add_Rot(&rotation);
		auto data = data_builder.Finish();

		m_Builder.Finish(data);

		*pLength = m_Builder.GetSize();

		return (char*)m_Builder.GetBufferPointer();
	}

	bool Rotation::Deserialize(const char* pChar)
	{
		auto data = flatbuffers::GetRoot<FBS::PlayerInput::Rotation>((const void*)pChar);

		m_nEntityID = data->EntityID();
		m_vec3Rotation.setValue(data->Rot()->x(), data->Rot()->y(), data->Rot()->z());

		return true;
	}
}

File: Client_Root/Client/Assets/Scripts/PlayerInput/PlayerInputs/Rotation.cs
using UnityEngine;
using FlatBuffers;

namespace PlayerInput
{
    public class Rotation : IPlayerInput
    {
        public int      m_nEntityID;
        public Vector3  m_vec3Rotation;

        public override FBS.PlayerInputType GetPlayerInputType()
        {
            return FBS.PlayerInputType.Rotation;
        }

        public override byte[] Serialize()
        {
            FBS.PlayerInput.Rotation.StartRotation(m_Builder);
            FBS.PlayerInput.Rotation.AddEntityID(m_Builder, m_nEntityID);
            FBS.PlayerInput.Rotation.AddRot(m_Builder, FBS.Data.Vector3.CreateVector3(m_Builder, m_vec3Rotation.x, m_vec3Rotation.y, m_vec3Rotation.z));
            var data = FBS.PlayerInput.Rotation.EndRotation(m_Builder);

            m_Builder.Finish(data.Value);

            return m_Builder.SizedByteArray();
        }

        public override bool Deserialize(byte[] bytes)
        {
            var buf = new ByteBuffer(bytes);

            var data = FBS.PlayerInput.Rotation.GetRootAsRotation(buf);

            m_nEntityID = data.EntityID;
            m_vec3Rotation = new Vector3(data.Rot.X, data.Rot.Y, data.Rot.Z);

            return true;
        }

        public override string ToString()
        {
            return string.Format("EntityID : {0}, Rotation : {1}", m_nEntityID, m_vec3Rotation.ToString());
        }
    }
}